File: domain/src/main/java/domain/Civilian.java
package domain;

import exceptions.KontoException;

import javax.persistence.*;
import java.io.Serializable;
import java.util.List;

@Entity
@DiscriminatorValue("Civilian")
public class Civilian extends User implements Serializable {

    private static final long serialVersionUID = 1L;

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    @ManyToMany
    private List<Vehicle> vehicles;
    @OneToMany
    private List<Invoice> invoices;

    public Civilian() {
        super();
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public List<Vehicle> getVehicles() {
        return vehicles;
    }

    public void setVehicles(List<Vehicle> vehicles) {
        this.vehicles = vehicles;
    }

    public void addVehicle(Vehicle vehicle) throws KontoException{
        if(!vehicles.contains(vehicle)){
            vehicles.add(vehicle);
        }
        throw new KontoException("user already owns car");
    }

    public void removeVehicle(Vehicle vehicle) throws KontoException {
        if(vehicles.contains(vehicle)){
            vehicles.remove(vehicle);
        }
        else{
            throw new KontoException("user does not own this car "+vehicle.getHardwareSerialNumber());
        }
    }

    public List<Invoice> getInvoices() {
        return invoices;
    }

    public void addInvoice(Invoice invoice) throws KontoException {
        if(!invoices.contains(invoice)) {
            invoices.add(invoice);
        }else{
            throw new KontoException("user already has invoice");
        }
    }

    public void setInvoices(List<Invoice> invoices) {
        this.invoices = invoices;
    }
}

File: src/main/java/domain/Civilian.kt
package domain

import java.io.Serializable
import java.util.* // ktlint-disable no-wildcard-imports
import javax.persistence.DiscriminatorValue
import javax.persistence.Entity
import javax.persistence.Id
import javax.persistence.ManyToMany
import javax.persistence.OneToMany

@Entity
@DiscriminatorValue("civilian")
class Civilian : KontoUser(), Serializable {

    @Id
    var id: String = UUID.randomUUID().toString()

    @ManyToMany
    var vehicles: List<Vehicle>? = null
    @OneToMany
    var invoices: List<Invoice>? = null

    companion object {

        private const val serialVersionUID = 1L
    }
}
